user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java

```diff
@@ -0,0 +1,101 @@
+import org.apache.kafka.clients.admin.ConsumerGroupDescription;
+import org.apache.kafka.common.ConsumerGroupState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.actuate.autoconfigure.health.ConditionalOnEnabledHealthIndicator;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.stereotype.Component;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Kafka health indicator. Check that the consumer group can be accessed and is registered with the cluster,
+ * if not mark the service as unhealthy.
+ */
+@Component("kafka")
+@ConditionalOnEnabledHealthIndicator("kafka")
+public class KafkaHealthIndicator implements HealthIndicator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaHealthIndicator.class);
+    private final AdminClient adminClient;
+    @Value("${akka.kafka.consumer.kafka-clients.group.id}")
+    private String groupId;
+
+    /**
+     * Requires the AdminClient to interact with Kafka
+     *
+     * @param adminClient of the cluster
+     */
+    public KafkaHealthIndicator(final AdminClient adminClient) {
+        this.adminClient = adminClient;
+    }
+
+    @Override
+    public Health getHealth(final boolean includeDetails) {
+        return Optional.of(performCheck())
+                .filter(healthy -> healthy)
+                .map(up -> Health.up().withDetail("group", this.groupId).build())
+                .orElseGet(() -> Health.down().withDetail("group", this.groupId).build());
+    }
+
+    /**
+     * Health endpoint
+     *
+     * @return the {@code Health} object
+     */
+    @Override
+    public Health health() {
+        return Optional.of(performCheck())
+                .filter(healthy -> healthy)
+                .map(up -> Health.up().build())
+                .orElseGet(() -> Health.down().build());
+    }
+
+    private boolean performCheck() {
+        try {
+            Map<String, ConsumerGroupDescription> groupDescriptionMap = this.adminClient.describeConsumerGroups(Collections.singletonList(this.groupId))
+                    .all()
+                    .get(1, TimeUnit.SECONDS);
+
+            ConsumerGroupDescription consumerGroupDescription = groupDescriptionMap.get(this.groupId);
+
+            LOGGER.debug("Kafka consumer group ({}) state: {}", groupId, consumerGroupDescription.state());
+
+            if (consumerGroupDescription.state() == ConsumerGroupState.STABLE) {
+                return consumerGroupDescription.members().stream()
+                        .noneMatch(member -> (member.assignment() == null || member.assignment().topicPartitions().isEmpty()));
```

Comment:
Would a LOGGER.error be good here if we fail and return false?

Reply:
In the case you return `false`, the service will soon be shut-down by k8s.
Any loggers here would get called periodically (every 20s or so?).
I think a `LOGGER.error("Failed to find kafka topic-partition assignments")` would be reasonable in this case.
Then a `kubectl logs user-service-0 -p` would reveal that there was a kafka problem quite quickly.

Reply:
updated to add a log message